### PR TITLE
Adding exist_ok=True in RAVEN parser to aviod FileExistsError

### LINKED ISSUE #1070: FileExistsError when making new folder
--------
Defect Description
--------

**Describe the defect**

##### What did you see instead?

When using raven creating folder, there might have a `FileExistsError`, because the `FileExistsError` is a new error type in python 3.

##### Do you have a suggested fix for the development team?

Adding `exist_ok=True` to os.makedirs

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by JiaZhou-PU**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
In RAVEN run RAVEN, it might have file exist error in method os.makedirs(slaveDir), closes #1070 

##### What are the significant changes in functionality due to this change request?

1.  Adding `exist_ok=True` inside os.makedirs() to avoid FileExistsError if a folder already exists, inside RAVEN parser and Steps.py.
2. Print warning messages if a folder already exists.
3. Make a new test for RAVEN run RAVEN to check the FileExistsError. This test will run twice, the first run will make subdirectories with 5 levels deep. The second run will proceed without clean the results from the previous run.
  a) Outer loop needs `inner_sample.xml`,`projectile.py` in as file inputs, the inner loop will need `proj_functions.py`  as the function for one variable, `input.i`(perturbable), and `inp.aux`(unperturbable) as the file input.
  b) When a copy of the raven input is made, whose `Files` node assigns an unperturbable file. The path will be modified as the absolute path of the original file.
  c) Made a copy of projectile.py to test folder, modified to error out if file not exist or sys.argv is not provide.

 

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by JiaZhou-PU** [comment]
@PaulTalbot-INL ready for review.

**Comment by PaulTalbot-INL** [comment]
This isn't limited to the HPC; it can randomly happen in parallel anywhere. It is definitely more common on the HPC.

@alfoa, I can't think a way to test this; the line is covered in existing tests, but I don't think we have a way to force a folder to get created between checking if it's there and making it. Is covering the line sufficient for testing?

**Comment by alfoa** [comment]
@JiaZhou-PU Did you talk to @PaulTalbot-INL ?

**Comment by PaulTalbot-INL** [comment]
Yes, I was wrong, it's not a parallel problem, it's a relative pathing problem.

For example, if I have the following structure:
```
  root/
      outer.xml
      inner.xml
      working_outer/
            step_name/
                  1/
                      inner.xml
                      working_inner/
```
and if I need the pickled ROM `arma.pk` in the inner, in the `inner.xml` inside `root/` I specify the path to the file as `../../arma.pk`, then when the OUTER tries to copy `root/working_inner/../../arma.pk` to `root/working_outer/step_name/1/working_inner/../../arma.pk`, of course this ends up being `root/working_outer/step_name/arma.pk` which the the same for each run. For every inner, the outer will try to rewrite this file.

This coincides with the problem that, really, `arma.pk` doesn't need to be copied at all, it can be referenced from a single location.

**Comment by alfoa** [comment]
Okay. So As I imagined, this PR does not fix anything. Right? we can close this and remove the issue.

**Comment by PaulTalbot-INL** [comment]
Well, it does fix the RAVENparser error, but not in the way we want to long term, as you suggest. Agreed.

**Comment by PaulTalbot-INL** [comment]
It looks like the root issue has been reported before: https://github.com/idaholab/raven/issues/753 but in older libraries the "copy" didn't error if the file already exists.

**Comment by JiaZhou-PU** [comment]
Check tests.

File: tests/framework/CodeInterfaceTests/RAVEN/Code/projectile.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  Input = {}
+  for line in open(inFile,'r'):
+    arg, val = (a.strip() for a in line.split('='))
+    if arg == 'aux':
+      auxFile = val
+    else:
+      Input[arg] = float(val)
+  # check auxfile exists
+  if not os.path.isfile(auxFile):
+    raise IOError('Aux file not found: {}'.format(auxFile))
```
Comment (JiaZhou-PU) [inline tests/framework/CodeInterfaceTests/RAVEN/Code/projectile.py]: Checking the sys.argv '-i -o' and the existence of aux file, only needed for this test

**Comment by JiaZhou-PU** [comment]
Ready for review @alfoa

File: tests/framework/CodeInterfaceTests/RAVEN/tests [hunk 2]
```diff
@@ -5,12 +5,28 @@
...
    max_time = 500
  [../]
+ [./Code]
+   type = 'RavenFramework'
+   input = 'code.xml'
+   UnorderedCsv = 'Code/outer_out.csv'
+   rel_err = 0.0001
+    [./folderExist]
+      type = Exists
+      output = 'Code/sample/1/innerRunDir/sample/4/out~input.csv '
```
Comment (JiaZhou-PU) [inline tests/framework/CodeInterfaceTests/RAVEN/tests]: Same input for the test, the first test will check the existence of the subdirectory of 5 levels deep.

File: tests/framework/CodeInterfaceTests/RAVEN/Code/projectile.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+#       v0 - initial total velocity
+#       ang - angle of initial motion, in degrees, with respect to flat ground
+#     Outputs:
+#       (x,y) - vector positions of projectile in time
+#       t - corresponding time steps
+#
+import os
+import numpy as np
+
+in_vars = ['x0', 'y0', 'v0', 'ang', 'timeOption']
```
Comment (alfoa) [inline tests/framework/CodeInterfaceTests/RAVEN/Code/projectile.py]: do we need to copy this here or we can just link it from the AnalyticalModel directory?

**Comment by alfoa** [comment]
Checklist passed.

Merging approved.

**Pull request merged**